### ISSUE TITLE
Interpolated + count aws_alb_target_group_attachment.target_id showing changes when increasing count or when reprovisioning a single target

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#253]
_This issue was originally opened by @levinse as hashicorp/terraform#8684. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Given this config

```
resource "aws_alb_target_group_attachment" "fe" {
  count = "${var.count}"

  target_group_arn = "${var.alb_target_group_https_arn}"
  target_id = "${element(aws_instance.fe.*.id, count.index)}"
  port = 8080
}

```

With count = 1 and changing to 2, I expect terraform plan to not show any changes to the existing first resource and to only show a change with the second resource added. However, terraform is showing the following plan output:

```
-/+ module.fe-app-stg.aws_alb_target_group_attachment.fe.0
    port:             "8080" => "8080"
    target_group_arn: "my ARN value" => "same ARN value"
    target_id:        "my instance id" => "${element(aws_instance.fe.*.id, count.index)}" (forces new resource)

```

Apparently this will destroy and re-create the first target group attachment when I add a second one - is that how the AWS API works or is that a bug?

**Comment by levinse** [comment]
@mitchellh any news on this? I've originally opened this bug a year ago...

Thanks

**Comment by aprevost** [comment]
I've been annoyed by the same thing... would love to see a fix.

**Comment by levinse** [comment]
Hello? Any news? We are happy to pay to get this fixed.

**Comment by austinmills** [comment]
I ran into this recently and it's extremely annoying -- basically resulting in unnecessary downtime unless you jump through a lot of hoops!

**Comment by thatrevguy** [comment]
Just ran into this same issue. Any word on when this will be fixed?

**Comment by dperetti** [comment]
Very annoying ! Defeats the purpose of load balancing...

**Comment by popopanda** [comment]
I ran into this issue, but I had a count variable for ec2 instance as well
```
resource "aws_instance" "ec2" {
  count                  = "${var.count}"
  ami                    = "${data.aws_ami.centos.id}"
  instance_type          = "${var.instance_type}"
....
}
```
```
resource "aws_lb_target_group_attachment" "apache" {
  count            = "${var.count}"
  target_group_arn = "${aws_lb_target_group.target_group.arn}"
  target_id        = "${element(module.apache.ec2_ids, count.index)}"
  port             = 443
}
```

When I increase the count variable to spin up additional ec2 instances, it would recreate the target_group_attachment, thus deregistering all the existing ec2 instances with the target_group, and re-adding them

```
[Truncated]
-/+ module.apache.aws_lb_target_group_attachment.apache[0] (new resource required)
-/+ module.apache.aws_lb_target_group_attachment.apache[1] (new resource required)
-/+ module.apache.aws_lb_target_group_attachment.apache[2] (new resource required)
-/+ module.apache.aws_lb_target_group_attachment.apache[3] (new resource required)
```

I was able to help work around by adding 

```
  lifecycle {
    ignore_changes = true
  }
```

to my `aws_lb_target_group_attachment` resource.

As a result, if I increase my count, my module creates the additional ec2 instance, as well as adding it to the target_group_attachment without affecting my existing ec2 instances part of the target_group

```
[Truncated]
+ module.apache.aws_lb_target_group_attachment.apache[4]

+ module.apache.module.apache.aws_instance.ec2[4]
```

I've tested scaling down (reducing my count variable), and it does this:
```
  - module.apache.aws_lb_target_group_attachment.apache[3]

  - module.apache.aws_lb_target_group_attachment.apache[4]

  - module.apache.module.apache.aws_instance.ec2[3]

  - module.apache.module.apache.aws_instance.ec2[4]

```

**Comment by bflad** [comment]
I provided some information in a pull request related to this issue about changes in the upcoming Terraform 0.12 release that will help with this situation: https://github.com/terraform-providers/terraform-provider-aws/pull/1726#issuecomment-418481437

**Comment by OferE** [comment]
this is a huge pain. When is this expected to be fixed?

**Comment by bflad** [comment]
Hi folks 👋 This issue is resolved in Terraform 0.12.6 and later, which supports new functionality in the configuration language aimed at solving problems like these. The new [resource-level `for_each` argument](https://www.terraform.io/docs/configuration/resources.html#for_each-multiple-resource-instances-defined-by-a-map-or-set-of-strings) can be used so resources are indexed in the Terraform state based on a string map or set, rather than the simple numeric list with the [resource-level `count` argument](https://www.terraform.io/docs/configuration/resources.html#count-multiple-resource-instances-by-count). Resources switched to `for_each` over `count` will no longer have issues with removing elements in the middle of a list or general rearranging of elements as the resource index keys are stable.

If you're looking for general assistance with how to implement `for_each` in this situation, please note that we use GitHub issues in this repository for tracking bugs and enhancements with the Terraform AWS Provider codebase rather than for questions. While we may be able to help with certain simple problems here it's generally better to use the [community forums](https://discuss.hashicorp.com/c/terraform-providers) where there are far more people ready to help, whereas the GitHub issues here are generally monitored only by [a few maintainers](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md#who-are-the-maintainers) and dedicated community members interested in code development of the Terraform AWS Provider itself.
